File: src/dev_health_ops/api/services/identity.py
"""Shared server-side identity resolution (Framework A7 / A8 / B7).

Resolves stable scope ids (``repo_id`` / ``team_id``) to human-readable
display names by querying the ``repos`` / ``teams`` analytics tables. This is
the same resolver contract the cockpit conclusion uses (CHAOS-2064) — extracted
here so Govern surfaces (Incident Correlation contributors, Change-Failure
associations, Coverage-by-Repository) can resolve identity server-side instead
of leaking ``#hex UNRESOLVED`` labels to the client.

Rules:
- Resolution happens server-side; ids are stable, display names are human.
- A bare UUID is never a valid display name (A8). Callers must apply a
  controlled fallback rather than surfacing the raw id.
- Best-effort: on any query failure an empty map is returned so callers fall
  back deterministically.
"""

from __future__ import annotations

import logging
import re
from typing import Any, Literal

from dev_health_ops.metrics.sinks.base import BaseMetricsSink

from ..queries.client import query_dicts

logger = logging.getLogger(__name__)

ScopeKind = Literal["repo", "team"]

_BARE_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)


def looks_like_uuid(value: str | None) -> bool:
    """Return True when *value* is a bare UUID that must not appear as a label (A8)."""
    if not value:
        return False
    return bool(_BARE_UUID_RE.match(value.strip()))


async def resolve_scope_display_names(
    sink: BaseMetricsSink,
    *,
    org_id: str,
    scope: ScopeKind,
    ids: list[str],
) -> dict[str, str]:
    """Return ``{scope_id: display_name}`` for the given ids.

    Queries the ``repos`` (``repo`` slug) or ``teams`` (``name``) table. Only
    non-UUID display names are kept — a row that would resolve to a bare UUID is
    omitted so callers apply a controlled fallback rather than surfacing the raw
    id (A8 / B7). Best-effort: returns an empty dict on any failure.
    """
    unique_ids = sorted({str(i) for i in ids if i})
    if not unique_ids:
        return {}

    if scope == "repo":
        sql = """
            SELECT toString(id) AS scope_id, repo AS display_name
            FROM repos
            WHERE org_id = {org_id:String}
              AND toString(id) IN {scope_ids:Array(String)}
        """
        params: dict[str, Any] = {"org_id": org_id, "scope_ids": unique_ids}
    else:
        sql = """
            SELECT toString(id) AS scope_id, name AS display_name
            FROM teams
            WHERE org_id = {org_id:String}
              AND toString(id) IN {scope_ids:Array(String)}
        """
        params = {"org_id": org_id, "scope_ids": unique_ids}

    try:
        rows = await query_dicts(sink, sql, params)
    except Exception:
        logger.warning(
            "Could not resolve %s display names for ids=%s", scope, unique_ids
        )
        return {}

    resolved: dict[str, str] = {}
    for row in rows:
        scope_id = str(row.get("scope_id") or "")
        display_name = row.get("display_name")
        if not scope_id or not display_name:
            continue
        display_name = str(display_name).strip()
        # A8: a bare UUID is not a human label — omit so caller falls back.
        if not display_name or looks_like_uuid(display_name):
            continue
        resolved[scope_id] = display_name
    return resolved


def scope_kind_for_group_by(group_by: str) -> ScopeKind:
    """Map an explain/driver ``group_by`` column to a resolver scope kind."""
    return "team" if group_by == "team_id" else "repo"

File: cmd/query-api/internal/explain/identity.go
package explain

import (
	"context"
	"fmt"
	"log"
	"regexp"
	"sort"
	"strings"

	dhclickhouse "github.com/full-chaos/dev-health-go/clickhouse"
)

// bareUUIDRe ports identity.py's _BARE_UUID_RE verbatim (case-insensitive,
// anchored both ends).
var bareUUIDRe = regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

// looksLikeUUID ports identity.py's looks_like_uuid.
func looksLikeUUID(value string) bool {
	v := strings.TrimSpace(value)
	if v == "" {
		return false
	}
	return bareUUIDRe.MatchString(v)
}

// scopeKindForGroupBy ports identity.py's scope_kind_for_group_by. Every
// metricConfigs entry's GroupBy is "team_id" or "repo_id" (metricconfig.go),
// so the "" fallback is unreached in practice, ported anyway for fidelity
// with the generic Python function.
func scopeKindForGroupBy(groupBy string) (string, bool) {
	switch groupBy {
	case "team_id":
		return "team", true
	case "repo_id":
		return "repo", true
	default:
		return "", false
	}
}

// resolveScopeDisplayNames ports resolve_scope_display_names
// (api/services/identity.py:45-99) -- best-effort: ANY failure (a bad
// query, a scan error, a row-iteration error) degrades to an empty map,
// never propagated as a route error, matching Python's own
// `except Exception: ... return {}`. Logged, not silently dropped.
//
// DEDUP: repos is ReplacingMergeTree(last_synced) (000_raw_tables.sql);
// teams is ReplacingMergeTree(updated_at) (002_teams.sql) -- BOTH since
// their original creation, not a later conversion. Python's own query
// here (identity.py:63-78) reads NEITHER table with FINAL or any argMax
// dedup -- a genuine ReplacingMergeTree-raw-read defect under the class
// ruling (an unmerged part can return more than one physical row for the
// same id, and/or a stale display_name for an id whose name changed).
// This port adds FINAL to both reads -- the Go answer is canonical (class
// ruling (b)); the difference from Python's raw read is a declared
// Python-plane defect, not a data-semantics choice.
func (reader *Reader) resolveScopeDisplayNames(ctx context.Context, orgID, scopeKind string, ids []string) map[string]string {
	unique := uniqueSortedNonEmpty(ids)
	if len(unique) == 0 {
		return map[string]string{}
	}

	var query string
	switch scopeKind {
	case "repo":
		query = fmt.Sprintf(`
SELECT toString(id) AS scope_id, repo AS display_name
FROM repos FINAL
WHERE org_id = {org_id:String}
  AND toString(id) IN {scope_ids:Array(String)}
%s
`, settingsMaxExecutionTime())
	case "team":
		query = fmt.Sprintf(`
SELECT toString(id) AS scope_id, name AS display_name
FROM teams FINAL
WHERE org_id = {org_id:String}
  AND toString(id) IN {scope_ids:Array(String)}
%s
`, settingsMaxExecutionTime())
	default:
		return map[string]string{}
	}

	if reader == nil || reader.client == nil {
		log.Printf("explain: resolve scope display names: client unavailable, scope=%s ids=%v", scopeKind, unique)
		return map[string]string{}
	}

	bindings := []dhclickhouse.Binding{
		{Name: "org_id", Value: orgID},
		{Name: "scope_ids", Value: unique},
	}
	rows, err := reader.client.Query(ctx, query, bindings)
	if err != nil {
		log.Printf("explain: could not resolve %s display names for ids=%v: %v", scopeKind, unique, err)
		return map[string]string{}
	}
	defer rows.Close()

	resolved := map[string]string{}
	for rows.Next() {
		var scopeID, displayName string
		if err := rows.Scan(&scopeID, &displayName); err != nil {
			log.Printf("explain: could not resolve %s display names for ids=%v: scan: %v", scopeKind, unique, err)
			return map[string]string{}
		}
		displayName = strings.TrimSpace(displayName)
		// A8: a bare UUID is not a human label -- omit so the caller
		// falls back to shortToken.
		if scopeID == "" || displayName == "" || looksLikeUUID(displayName) {
			continue
		}
		resolved[scopeID] = displayName
	}
	if err := rows.Err(); err != nil {
		log.Printf("explain: could not resolve %s display names for ids=%v: iterate: %v", scopeKind, unique, err)
		return map[string]string{}
	}
	return resolved
}

func uniqueSortedNonEmpty(ids []string) []string {
	seen := make(map[string]struct{}, len(ids))
	var out []string
	for _, id := range ids {
		if id == "" {
			continue
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		out = append(out, id)
	}
	sort.Strings(out)
	return out
}
